connection-loop.ts

```diff
@@ -0,0 +1,286 @@
+let pushCounter = 0;
+const MIN_DELAY = 30;
+const MAX_DELAY = 60_000;
+
+const MAX_CONNECTIONS = 3;
+
+// Number of connections to remember when computing the new delay.
+const CONNECTION_MEMORY_COUNT = 10;
+
+type Duration = {start: number; stop: number; ok: boolean};
+type Durations = Duration[];
+
+const PUSH_DELAY = 10;
+
+class Replicache {
+  private _pendingPushResolver = resolver<void>();
+  private _currentPushResolver = resolver<void>();
+  private _closed = false;
+
+  async push() {
+    this._pendingPushResolver.resolve();
+    // await a promise to allow the currentPushResolver to be created inside the
+    // connection loop.
+    await Promise.resolve(0);
+    await this._currentPushResolver.promise;
+  }
+
+  async connectionPushLoop() {
+    const durations: Durations = [];
+    let delay = MIN_DELAY;
+    let lastConnectionErrored = false;
+    console.log('Starting connection loop');
+
+    while (!this._closed) {
+      console.log('Waiting for a push');
+      await this._pendingPushResolver.promise;
+
+      // This resolvers is used to make the individual calls to push have the
+      // correct "duration".
+      const currentPushResolver = resolver();
+      this._currentPushResolver = currentPushResolver;
+      currentPushResolver.promise.catch(() => 0);
+
+      await sleep(PUSH_DELAY);
+
+      // This resolver is used to wait for incoming push calls.
+      this._pendingPushResolver = resolver();
+
+      if (pushCounter >= MAX_CONNECTIONS) {
+        console.log('Too many pushes. Waiting until one finishes');
+        await waitUntilAvailableConnection();
+      }
+
+      if (lastConnectionErrored || pushCounter > 0) {
```

Comment:
hm. I do not see the bit where a push gets retried if it fails. Like it will get picked up again implicity if somebody calls push again, but maybe won't otherwise?

We delay below in the case of error, but that doesn't by itself cause us to retry the request does it?

Reply:
Checking...

Reply:
API Q: Does a call to `push()` that fails reject? or do we just not resolve it until a successful push is done?